src/webui/api/torrentscontroller.cpp

```diff
@@ -1205,6 +1205,27 @@ void TorrentsController::setAutoManagementAction()
     });
 }
 
+void TorrentsController::setSSLCertificateAction()
+{
+    requireParams({u"hash"_s});
+    const auto id = BitTorrent::TorrentID::fromString(params()[u"hash"_s]);
+    BitTorrent::Torrent *const torrent = BitTorrent::Session::instance()->getTorrent(id);
+    if (!torrent)
+        throw APIError(APIErrorType::NotFound);
+
+    for (const auto &file: {u"certificate"_s, u"privateKey"_s, u"dhParams"_s})
+    {
+        if (!data().contains(file))
+            throw APIError(APIErrorType::BadData);
+    }
+
+    const QByteArray certificate = data()[u"certificate"_s];
+    const QByteArray privateKey = data()[u"privateKey"_s];
+    const QByteArray dhParams = data()[u"dhParams"_s];
+
+    torrent->setSSLCertificate(certificate, privateKey, dhParams);
```

Comment:
FWIW we had a similar situation with the webui certificate. A long time ago it was the same as this, importing the data directly. However later it became a problem when the certificates expired and users need to reload them. It was suggested that qbt should record the cert file paths instead. This way users can just replace the cert files and restart the webui for it to take effect. I suppose we should follow that lesson here.
Also the folder/file creation in `TorrentImpl::setSSLCertificate()` (and therefore `m_sslCertificatesDirectory` too) seems to be doing unrelated things (unrelated to torrenting) and I would like to somehow simplify it. For example, one would have guessed that `TorrentImpl::setSSLCertificate` might be reading cert files and passing the data to libtorrent instead of saving certs to disk.

Reply:
Excuse me I wasn't following the latest development and discussions.

>The certificates from that path will be automatically added to affected torrents when torrent_need_cert_alert is raised. 

IMO for every torrent the cert paths would need be designated separately. And of course we can have some default paths for every non-specified torrent and users can opt to use it if they want.

Reply:
I pushed the change to set certificates using their path. But I don't fully understand the suggestion about deporting the logic for creation of directories/certificates somewhere else. Would it be better located in the m_session? On the last subject: Is the configurable per-torrent certificate path something difficult to add later ? Or is it desired in this initial implementation ?

Reply:
>But I don't fully understand the suggestion about deporting the logic for creation of directories/certificates somewhere else. Would it be better located in the m_session? 

Just to be clear, I was suggesting that qbt don't handle the creation of directories/certificates. Instead qbt will just take cert paths from user and pass it to libtorrent. The cert will need to be placed on disk beforehand and this part is done manually by the user.
Also, regarding security, I suppose this is better since the certs aren't transmitted over the net and therefore we don't need top security for our web server/api implementation.

>I pushed the change to set certificates using their path. 

Not sure if you meant it is done. But I would expect more changes if that idea of "pass cert path" is accepted. I.e. the `setSSLCertificateAction` will not accept cert data and only accept cert paths. (and rename it to `setSSLCertificatePathAction`?)

>On the last subject: Is the configurable per-torrent certificate path something difficult to add later ? Or is it desired in this initial implementation ?

I don't think it will be too difficult:
```c++
void TorrentImpl::setSSLCertificate(const Path &certificatePath, const Path &privateKeyPath, const Path &dhParamsPath)
{
    // if required, check cert validness of certificatePath, privateKeyPath, dhParamsPath and log error if any
    m_nativeHandle.set_ssl_certificate(certificatePath.toString(), privateKeyPath.toString(), dhParamsPath.toString());
}
```
And I don't insist the UI side to be complete in this PR.

Reply:
I'm personally interested in the workflow where the actual certificate content is set via the API. 

It doesn't prevent us from implementing two separate actions: one for setting the certificates via path; the other using their content, but I would really like to see a variation of the current behavior implemented.

Reply:
Also, note that the code snippet which you proposed : 
```
void TorrentImpl::setSSLCertificate(const Path &certificatePath, const Path &privateKeyPath, const Path &dhParamsPath)
{
    // if required, check cert validness of certificatePath, privateKeyPath, dhParamsPath and log error if any
    m_nativeHandle.set_ssl_certificate(certificatePath.toString(), privateKeyPath.toString(), dhParamsPath.toString());
}
``` 
doesn't encounter for the fact that m_nativeHandle doesn't preserve the configuration across restarts. So we'll need to persist the paths passed to this function on qbittorrent side. On restart, libtorrent raises the `handleTorrentNeedCertAlert` alert and qbittorrent would have to call `m_nativeHandle.set_ssl_certificate` again with the paths it took care of preserving accross restarts.

Reply:
As for setting certificates via paths, does it really make sense to make it more complicated than just limiting yourself to use pre-defined paths already implemented in this PR? I.e. `<configured_certificates_for>/<info_hash>.<ext>`. Otherwise it would require to implement the way of assigning/changing and storing such custom paths within torrent "resume" data.

Reply:
Excuse my late reply, was busy in weekdays.

>the other using their content

Personally I don't like this way. It doesn't scale well if the certs are many and qbt will be responsible for storing/managing them.

>So we'll need to persist the paths passed to this function on qbittorrent side. On restart, libtorrent raises the handleTorrentNeedCertAlert alert and qbittorrent would have to call m_nativeHandle.set_ssl_certificate again with the paths it took care of preserving accross restarts.

In an ideal world, the paths would be stored in torrent resume data (and able to be manipulated in UI/API) like glassez said. However I don't oppose a 'simpler' implementation like glassez [suggestion](https://github.com/qbittorrent/qBittorrent/pull/19542#discussion_r1384346903) for now.

Reply:
I clearly miss-understand the scalability issue which worries you. Can you please be more specific in the explanation of your concerns?

Reply:
@Chocobo1 , I allow myself to re-iterate my question from the previous message: what are the scalability concerns for the current solution ?

Reply:
>@Chocobo1 , I allow myself to re-iterate my question from the previous message: what are the scalability concerns for the current solution ?

Sorry for (very) late reply. I intend get my hands on this PR in the near future. Maybe I can clear things up when that time comes.

Reply:
I believe that we should imagine a typical scenario of its use, so that in the end it does not turn out that we get a very reliable, scalable, etc., but at the same time a unusable feature.

Reply:
@rcarpa 
I have played with SSL torrents a bit and in the mean time, had a little discussion with other team members. I understand SSL torrents better than before. Also I had a bit different idea/approach to support SSL torrents in qbt. I figure it would be easier for us that I submit my ideas in another PR for further discussions: #20338